File: jetbrains/src/main/kotlin/deplens/lang/java/GradleDepLensInlayProvider.kt
package deplens.lang.java

import com.intellij.codeInsight.hints.declarative.InlayTreeSink
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import deplens.lang.BaseDepLensInlayProvider
import deplens.utils.inlay.GithubInlayUtils
import deplens.utils.resolver.MavenRepoResolver

class GradleDepLensInlayProvider : BaseDepLensInlayProvider() {

    private val depNotationRegex = Regex(
        """(implementation|api|compileOnly|runtimeOnly|testImplementation|testCompileOnly|testRuntimeOnly|kapt|annotationProcessor|classpath|compile|testCompile)\s*\(?\s*['\"]([A-Za-z0-9_.-]+):([A-Za-z0-9_.-]+):([A-Za-z0-9_.-]+)[^'\"\s\)]*['\"]"""
    )

    private val mapNotationRegex = Regex(
        """(implementation|api|compileOnly|runtimeOnly|testImplementation|testCompileOnly|testRuntimeOnly|kapt|annotationProcessor|classpath|compile|testCompile)\s*\(?\s*group:\s*['\"]([A-Za-z0-9_.-]+)['\"]\s*,\s*name:\s*['\"]([A-Za-z0-9_.-]+)['\"](?:\s*,\s*version:\s*['\"]([A-Za-z0-9_.-]+)['\"])?"""
    )

    override fun isFileSupported(file: PsiFile): Boolean {
        val name = file.name
        return name == "build.gradle" || name == "build.gradle.kts"
    }

    override fun collectElement(file: PsiFile, element: PsiElement, sink: InlayTreeSink) {
        if (element !is PsiFile) return

        val text = element.text

        for (match in depNotationRegex.findAll(text)) {
            val groupId = match.groupValues[2]
            val artifactId = match.groupValues[3]
            val version = match.groupValues[4]
            val repoKey = MavenRepoResolver.repoKeyFromGroupArtifact(groupId, artifactId, version) ?: continue
            val offset = match.range.last + 1
            GithubInlayUtils.addRepoInlay(file, sink, repoKey, offset)
        }

        for (match in mapNotationRegex.findAll(text)) {
            val groupId = match.groupValues[2]
            val artifactId = match.groupValues[3]
            val version = match.groupValues.getOrNull(4)
            val repoKey = MavenRepoResolver.repoKeyFromGroupArtifact(groupId, artifactId, version) ?: continue
            val offset = match.range.last + 1
            GithubInlayUtils.addRepoInlay(file, sink, repoKey, offset)
        }
    }
}

File: vscode/src/lang/java/GradleDepLensInlayProvider.ts
import type * as vscode from "vscode";
import { GithubInlayUtils } from "../../utils/inlay/GithubInlayUtils";
import { MavenRepoResolver } from "../../utils/resolver/MavenRepoResolver";
import { GithubRepoInfoService } from "../../utils/service/GithubRepoInfoService";
import { BaseDepLensInlayProvider } from "../BaseDepLensInlayProvider";

const DEP_NOTATION_REGEX =
  /(implementation|api|compileOnly|runtimeOnly|testImplementation|testCompileOnly|testRuntimeOnly|kapt|annotationProcessor|classpath|compile|testCompile)\s*\(?\s*['"]([A-Za-z0-9_.-]+):([A-Za-z0-9_.-]+):([A-Za-z0-9_.\-+$]+)[^'"\s\)]*['"]/g;
const MAP_NOTATION_REGEX =
  /(implementation|api|compileOnly|runtimeOnly|testImplementation|testCompileOnly|testRuntimeOnly|kapt|annotationProcessor|classpath|compile|testCompile)\s*\(?\s*group:\s*['"]([A-Za-z0-9_.-]+)['"]\s*,\s*name:\s*['"]([A-Za-z0-9_.-]+)['"](?:\s*,\s*version:\s*['"]([^'"]+)['"])?/g;

export class GradleDepLensInlayProvider extends BaseDepLensInlayProvider {
  constructor() {
    super();
    GithubRepoInfoService.getInstance().onDidUpdateRepoInfo(() => {
      this.emitter.fire();
    });
  }

  protected isFileSupported(document: vscode.TextDocument): boolean {
    return (
      document.fileName.endsWith("build.gradle") || document.fileName.endsWith("build.gradle.kts")
    );
  }

  protected async provideInlayHintsForDocument(
    document: vscode.TextDocument,
    range: vscode.Range,
    token: vscode.CancellationToken,
  ): Promise<vscode.InlayHint[]> {
    const hints: vscode.InlayHint[] = [];
    const text = document.getText();

    await this.collectMatches(text, document, range, hints, DEP_NOTATION_REGEX, 2, 3, 4);
    await this.collectMatches(text, document, range, hints, MAP_NOTATION_REGEX, 2, 3, 4);

    return hints;
  }

  private async collectMatches(
    text: string,
    document: vscode.TextDocument,
    range: vscode.Range,
    hints: vscode.InlayHint[],
    regex: RegExp,
    groupIndex: number,
    artifactIndex: number,
    versionIndex: number,
  ): Promise<void> {
    for (const match of text.matchAll(regex)) {
      if (match.index === undefined) {
        continue;
      }

      const position = document.positionAt(match.index + match[0].length);
      if (!range.contains(position)) {
        continue;
      }

      const repoKey = await MavenRepoResolver.repoKeyFromGroupArtifact(
        match[groupIndex],
        match[artifactIndex],
        match[versionIndex],
      );
      if (!repoKey) {
        continue;
      }

      GithubInlayUtils.addRepoInlay(hints, position, repoKey);
    }
  }
}
